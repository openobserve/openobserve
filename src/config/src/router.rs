// Copyright 2025 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

const QUERIER_ROUTES: [&str; 20] = [
    "/config",
    "/summary",
    "/organizations",
    "/settings",
    "/schema",
    "/streams",
    "/clusters",
    "/query_manager",
    "/ws",
    "/_search",
    "/_around",
    "/_values",
    "/functions?page_num=",
    "/prometheus/api/v1/series",
    "/prometheus/api/v1/query",
    "/prometheus/api/v1/query_range",
    "/prometheus/api/v1/query_exemplars",
    "/prometheus/api/v1/metadata",
    "/prometheus/api/v1/labels",
    "/prometheus/api/v1/label/",
];
const QUERIER_ROUTES_BY_BODY: [&str; 2] = [
    "/prometheus/api/v1/query_range",
    "/prometheus/api/v1/query_exemplars",
];
const FIXED_QUERIER_ROUTES: [&str; 3] = ["/summary", "/schema", "/streams"];

#[inline]
pub fn is_querier_route(path: &str) -> bool {
    QUERIER_ROUTES.iter().any(|x| path.contains(x))
}

#[inline]
pub fn is_querier_route_by_body(path: &str) -> bool {
    QUERIER_ROUTES_BY_BODY.iter().any(|x| path.contains(x))
}

#[inline]
pub fn is_fixed_querier_route(path: &str) -> bool {
    FIXED_QUERIER_ROUTES.iter().any(|x| path.contains(x))
}
