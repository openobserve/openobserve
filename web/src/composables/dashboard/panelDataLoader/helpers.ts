// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { useStore } from "vuex";
import { isWebSocketEnabled, isStreamingEnabled } from "@/utils/zincutils";
import type { PanelState } from "./types";
import { PANEL_DATA_LOADER_DEBOUNCE_TIME } from "./constants";

export class PanelHelpers {
  private store = useStore();

  constructor(
    private state: PanelState,
    private panelSchema: any,
  ) {}

  shouldFetchAnnotations() {
    return [
      "area",
      "area-stacked",
      "bar",
      "h-bar",
      "line",
      "scatter",
      "stacked",
      "h-stacked",
    ].includes(this.panelSchema.value.type);
  }

  hasAtLeastOneQuery() {
    return this.panelSchema.value.queries?.some((q: any) => q?.query);
  }

  waitForTimeout(signal: AbortSignal) {
    return new Promise<void>((resolve, reject) => {
      const timeoutId = setTimeout(resolve, PANEL_DATA_LOADER_DEBOUNCE_TIME);

      signal.addEventListener("abort", () => {
        clearTimeout(timeoutId);
        reject(new Error("Aborted waiting for loading"));
      });
    });
  }

  waitForThePanelToBecomeVisible(signal: any, isVisible: any, forceLoad: any) {
    return new Promise<void>((resolve, reject) => {
      if (forceLoad.value == true) {
        resolve();
        return;
      }
      
      if (isVisible.value) {
        resolve();
        return;
      }

      const stopWatching = () => {}; // This will be implemented in main composable
      
      signal.addEventListener("abort", () => {
        stopWatching();
        reject(new Error("Aborted waiting for loading"));
      });
    });
  }

  addTraceId(traceId: string) {
    if (this.state.searchRequestTraceIds.includes(traceId)) {
      return;
    }
    this.state.searchRequestTraceIds = [...this.state.searchRequestTraceIds, traceId];
  }

  removeTraceId(traceId: string) {
    this.state.searchRequestTraceIds = this.state.searchRequestTraceIds.filter(
      (id: any) => id !== traceId,
    );
  }

  processApiError(error: any, type: any) {
    switch (type) {
      case "promql": {
        const errorDetailValue = error?.response?.data?.error || error?.message;
        const trimmedErrorMessage =
          errorDetailValue?.length > 300
            ? errorDetailValue.slice(0, 300) + " ..."
            : errorDetailValue;

        const errorCode =
          error?.response?.status ||
          error?.status ||
          error?.response?.data?.code ||
          "";

        this.state.errorDetail = {
          message: trimmedErrorMessage,
          code: errorCode,
        };
        break;
      }
      case "sql": {
        const errorDetailValue =
          error?.response?.data.error_detail ||
          error?.response?.data.message ||
          error?.error_detail ||
          error?.message ||
          error?.error;

        const trimmedErrorMessage =
          errorDetailValue?.length > 300
            ? errorDetailValue.slice(0, 300) + " ..."
            : errorDetailValue;

        const errorCode =
          isWebSocketEnabled(this.store.state) || isStreamingEnabled(this.store.state)
            ? error?.response?.status ||
              error?.status ||
              error?.response?.data?.code ||
              error?.code ||
              ""
            : error?.response?.status ||
              error?.response?.data?.code ||
              error?.status ||
              error?.code ||
              "";

        this.state.errorDetail = {
          message: trimmedErrorMessage,
          code: errorCode,
        };
        break;
      }
      default:
        break;
    }
  }
}
