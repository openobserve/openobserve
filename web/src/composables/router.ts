// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import Login from "@/views/Login.vue";
import { useLocalUserInfo, useLocalCurrentUser } from "@/utils/zincutils";
import useIngestionRoutes from "./shared/useIngestionRoutes";

const useOSRoutes = () => {
  const parentRoutes = [
    {
      path: "/login",
      component: Login,
    },
    {
      path: "/logout",
      beforeEnter(to: any, from: any, next: any) {
        useLocalCurrentUser("", true);
        useLocalUserInfo("", true);

        window.location.href = "/login";
      },
    },
    {
      path: "/cb",
      name: "callback",
      component: Login,
    },
  ];

  const homeChildRoutes: any[] = [...useIngestionRoutes()];

  return { parentRoutes, homeChildRoutes };
};

export default useOSRoutes;
