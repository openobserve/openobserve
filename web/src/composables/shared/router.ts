// Copyright 2023 Zinc Labs Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import Home from "@/views/HomeView.vue";
import Tickets from "@/views/TicketsView.vue";
import About from "@/views/About.vue";
const ViewDashboard = () => import("@/views/Dashboards/ViewDashboard.vue");
const AddPanel = () => import("@/views/Dashboards/addPanel/AddPanel.vue");
import MemberSubscription from "@/views/MemberSubscription.vue";
const Search = () => import("@/views/Search.vue");
const AppMetrics = () => import("@/views/AppMetrics.vue");
const AppTraces = () => import("@/views/AppTraces.vue");
const LogStream = () => import("@/views/LogStream.vue");
const StreamExplorer = () => import("@/views/StreamExplorer.vue");
const Alerts = () => import("@/views/AppAlerts.vue");
import Error404 from "@/views/Error404.vue";
const Dashboards = () => import("@/views/Dashboards/Dashboards.vue");
const AlertList = () => import("@/components/alerts/AlertList.vue");
const TemplateList = () => import("@/components/alerts/TemplateList.vue");
const DestinationList = () => import("@/components/alerts/DestinationList.vue");
const Settings = () => import("@/components/settings/index.vue");

import ImportDashboard from "@/views/Dashboards/ImportDashboard.vue";
const Functions = () => import("@/views/Functions.vue");
const FunctionList = () => import("@/components/functions/FunctionList.vue");
const AssociatedStreamFunction = () =>
  import("@/components/functions/AssociatedStreamFunction.vue");
const EnrichmentTableList = () =>
  import("@/components/functions/EnrichmentTableList.vue");
const RealUserMonitoring = () => import("@/views/RUM/RealUserMonitoring.vue");
const SessionViewer = () => import("@/views/RUM/SessionViewer.vue");
const ErrorViewer = () => import("@/views/RUM/ErrorViewer.vue");
const AppPerformance = () => import("@/views/RUM/AppPerformance.vue");
const AppErrors = () => import("@/views/RUM/AppErrors.vue");
const AppSessions = () => import("@/views/RUM/AppSessions.vue");

const ReportList = () => import("@/components/reports/ReportList.vue");
const CreateReport = () => import("@/components/reports/CreateReport.vue");

const PerformanceSummary = () =>
  import("@/components/rum/performance/PerformanceSummary.vue");
const WebVitalsDashboard = () =>
  import("@/components/rum/performance/WebVitalsDashboard.vue");
const ErrorsDashboard = () =>
  import("@/components/rum/performance/ErrorsDashboard.vue");
const ApiDashboard = () =>
  import("@/components/rum/performance/ApiDashboard.vue");
const StreamRouting = () => import("@/components/functions/StreamRouting.vue");
const PipelineEditor = () => import("@/components/pipeline/PipelineEditor.vue");
const PipelinesList = () => import("@/components/pipeline/PipelinesList.vue");

import { routeGuard } from "@/utils/zincutils";
import useIngestionRoutes from "./useIngestionRoutes";
import useIamRoutes from "./useIamRoutes";
import config from "@/aws-exports";
import useManagementRoutes from "./useManagementRoutes";

const useRoutes = () => {
  const parentRoutes: never[] = [];

  const homeChildRoutes = [
    {
      path: "",
      name: "home",
      component: Home,
      meta: {
        keepAlive: true,
      },
    },
    {
      path: "logs",
      name: "logs",
      component: Search,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "metrics",
      name: "metrics",
      component: AppMetrics,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "traces",
      name: "traces",
      component: AppTraces,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      name: "streamExplorer",
      path: "streams/stream-explore",
      component: StreamExplorer,
      props: true,
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "streams",
      name: "logstreams",
      component: LogStream,
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "tickets",
      name: "tickets",
      component: Tickets,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "about",
      name: "about",
      component: About,
      meta: {
        keepAlive: true,
      },
    },
    {
      path: "dashboards",
      name: "dashboards",
      component: Dashboards,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "/dashboards/view",
      name: "viewDashboard",
      component: ViewDashboard,
      props: true,
      // meta: {
        // keepAlive: true,
      // },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "/dashboards/import",
      name: "importDashboard",
      component: ImportDashboard,
      props: true,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "/dashboards/add_panel",
      name: "addPanel",
      component: AddPanel,
      props: true,
      meta: {
        // keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    {
      path: "member_subscription",
      name: "member_subscription",
      component: MemberSubscription,
      meta: {
        keepAlive: true,
      },
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
    },
    ...useManagementRoutes(),
    {
      path: "pipeline",
      name: "pipeline",
      component: Functions,
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
      children: [
        {
          path: "functions",
          name: "functionList",
          component: FunctionList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "stream-association",
          name: "streamFunctions",
          component: AssociatedStreamFunction,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "enrichment-tables",
          name: "enrichmentTables",
          component: EnrichmentTableList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "pipelines",
          name: "pipelines",
          component: PipelinesList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
          children: [
            {
              path: "edit",
              name: "pipelineEditor",
              component: PipelineEditor,
              beforeEnter(to: any, from: any, next: any) {
                routeGuard(to, from, next);
              },
            },
          ],
        },
      ],
    },
    {
      path: "alerts",
      name: "alerts",
      component: Alerts,
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
      children: [
        {
          path: "alerts",
          name: "alertList",
          component: AlertList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "destinations",
          name: "alertDestinations",
          component: DestinationList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "templates",
          name: "alertTemplates",
          component: TemplateList,
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
      ],
    },
    {
      path: "rum",
      name: "RUM",
      component: RealUserMonitoring,
      beforeEnter(to: any, from: any, next: any) {
        routeGuard(to, from, next);
      },
      children: [
        {
          path: "sessions",
          name: "Sessions",
          component: AppSessions,
          meta: {
            keepAlive: true,
          },
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "sessions/view/:id",
          name: "SessionViewer",
          component: SessionViewer,
          props: true,
          meta: {
            keepAlive: false,
          },
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "errors",
          name: "ErrorTracking",
          component: AppErrors,
          meta: {
            keepAlive: true,
          },
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "errors/view/:id",
          name: "ErrorViewer",
          component: ErrorViewer,
          props: true,
          meta: {
            keepAlive: true,
          },
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
        },
        {
          path: "performance",
          name: "RumPerformance",
          component: AppPerformance,
          props: true,
          meta: {
            keepAlive: true,
          },
          beforeEnter(to: any, from: any, next: any) {
            routeGuard(to, from, next);
          },
          children: [
            {
              path: "overview",
              name: "rumPerformanceSummary",
              component: PerformanceSummary,
              beforeEnter(to: any, from: any, next: any) {
                routeGuard(to, from, next);
              },
            },
            {
              path: "web-vitals",
              name: "rumPerformanceWebVitals",
              component: WebVitalsDashboard,
              beforeEnter(to: any, from: any, next: any) {
                routeGuard(to, from, next);
              },
            },
            {
              path: "errors",
              name: "rumPerformanceErrors",
              component: ErrorsDashboard,
              beforeEnter(to: any, from: any, next: any) {
                routeGuard(to, from, next);
              },
            },
            {
              path: "apis",
              name: "rumPerformanceApis",
              component: ApiDashboard,
              beforeEnter(to: any, from: any, next: any) {
                routeGuard(to, from, next);
              },
            },
          ],
        },
      ],
    },
    ...useIngestionRoutes(),
    ...useIamRoutes(),
    {
      path: "/:catchAll(.*)*",
      component: Error404,
      meta: {
        keepAlive: true,
      },
    },
  ];

  if (config.isCloud === "false") {
    homeChildRoutes.splice(
      13,
      0,
      {
        path: "/reports",
        name: "reports",
        component: ReportList,
        props: true,
        beforeEnter(to: any, from: any, next: any) {
          routeGuard(to, from, next);
        },
      },
      {
        path: "/reports/create",
        name: "createReport",
        component: CreateReport,
        props: true,
        beforeEnter(to: any, from: any, next: any) {
          routeGuard(to, from, next);
        },
      }
    );
  }

  return { parentRoutes, homeChildRoutes };
};

export default useRoutes;
