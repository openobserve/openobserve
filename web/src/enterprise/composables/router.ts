// Copyright 2023 Zinc Labs Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import Login from "@/views/Login.vue";
import LoginCallback from "@/enterprise/components/login/Login.vue";
import {
  useLocalUserInfo,
  useLocalCurrentUser,
  getLoginURL,
  getLogoutURL,
  invlidateLoginData,
} from "@/utils/zincutils";

import authService from "@/services/auth";
import config from "@/aws-exports";
import Organizations from "@/enterprise/components/organizations/Organization.vue";
import Billing from "@/enterprise/components/billings/Billing.vue";
import Plans from "@/enterprise/components/billings/plans.vue";
import InvoiceHistory from "@/enterprise/components/billings/invoiceHistory.vue";
import Usage from "@/enterprise/components/billings/usage.vue";
import { routeGuard } from "@/utils/zincutils";

const useEnvRoutes = () => {
  const parentRoutes = [
    {
      path: "/login",
      component: Login,
      beforeEnter: async (to: any, from: any, next: any) => {
        if (config.isEnterprise == "true") {
          try {
            const url = await authService.get_dex_login();
            if (url) {
              window.location.href = url;
            } else {
              next();
            }
          } catch (error) {
            console.error("Error during redirection:", error);
            next(false);
          }
        } else {
          window.location.href = getLoginURL();
        }
      },
    },
    {
      path: "/logout",
      beforeEnter: async (to: any, from: any, next: any) => {
        useLocalCurrentUser("", true);
        useLocalUserInfo("", true);
        if (config.isEnterprise == "true") {
          invlidateLoginData();
        }
        //  else {
        //   window.location.href = getLogoutURL();
        // }
      },
    },
    {
      path: "/cb",
      name: "callback",
      component: LoginCallback,
    },
  ];

  const homeChildRoutes = [
    {
      path: "billings",
      name: "billings",
      component: Billing,
      meta: {
        keepAlive: false,
      },
      children: [
        {
          path: "usage",
          name: "usage",
          component: Usage,
        },
        {
          path: "plans",
          name: "plans",
          component: Plans,
        },
        {
          path: "invoice_history",
          name: "invoice_history",
          component: InvoiceHistory,
        },
      ],
    },
  ];

  return { parentRoutes, homeChildRoutes };
};

export default useEnvRoutes;
