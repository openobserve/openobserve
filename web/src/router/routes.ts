// Copyright 2023 Zinc Labs Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { createRouter, createWebHistory } from "vue-router";
import { getPath, mergeRoutes } from "@/utils/zincutils";

import userRoutes from "@/composables/shared/router";
import useOSRoutes from "@/composables/router";
let { parentRoutes, homeChildRoutes } = userRoutes();

const envRoutes: any = useOSRoutes();

parentRoutes = mergeRoutes(parentRoutes, envRoutes.parentRoutes);
homeChildRoutes = mergeRoutes(homeChildRoutes, envRoutes.homeChildRoutes);

const routes = [
  ...parentRoutes,
  {
    path: "/",
    component: () => import("@/layouts/MainLayout.vue"),
    children: [...homeChildRoutes],
  },
];
interface RouterMap {
  history: any;
  routes: any;
}
const routerMap: RouterMap = {
  history: createWebHistory(getPath()),
  // history: createWebHistory(window.location.pathname),
  routes: routes,
};

export default createRouter(routerMap);
