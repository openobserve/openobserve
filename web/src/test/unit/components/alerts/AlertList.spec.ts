// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { describe, expect, it, beforeEach, vi, afterEach } from "vitest";
import { mount, flushPromises, DOMWrapper } from "@vue/test-utils";
import { installQuasar } from "../../helpers/install-quasar-plugin";
import { Dialog, Notify } from "quasar";
import AlertList from "@/components/alerts/AlertList.vue";
import i18n from "@/locales";
import store from "../../helpers/store";
// @ts-ignore
import { rest } from "msw";
import AlertService from "@/services/alerts";
import router from "../../helpers/router";

const node = document.createElement("div");
node.setAttribute("id", "app");
document.body.appendChild(node);

installQuasar({
  plugins: [Dialog, Notify],
});

describe("Alert List", async () => {
  let wrapper: any;
  beforeEach(async () => {
    vi.useFakeTimers();
    wrapper = mount(AlertList, {
      attachTo: "#app",
      global: {
        provide: {
          store: store,
        },
        plugins: [i18n, router],
      },
    });
    await flushPromises();
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("Should render alerts title", () => {
    expect(wrapper.find('[data-test="alerts-list-title"]').text()).toBe(
      "Alerts"
    );
  });
  it("Should reder table with alerts", () => {
    expect(
      wrapper.find('[data-test="alert-list-table"]').exists()
    ).toBeTruthy();
  });

  it("Should display table column headers", async () => {
    await flushPromises();
    const tableData = wrapper
      .find('[data-test="alert-list-table"]')
      .find("thead")
      .find("tr")
      .findAll("th");
    expect(tableData[0].text()).toBe("#");
    expect(tableData[1].text()).toContain("Name");
    expect(tableData[2].text()).toContain("Stream Type");
    expect(tableData[3].text()).toContain("Stream Name");
    expect(tableData[4].text()).toContain("Query");
    expect(tableData[5].text()).toContain("Condition");
    expect(tableData[6].text()).toContain("Destination");
    expect(tableData[7].text()).toContain("Actions");
  });

  it("Should display table row data", async () => {
    await flushPromises();
    const tableData = wrapper
      .find('[data-test="alert-list-table"]')
      .find("tbody")
      .find("tr")
      .findAll("td");
    expect(tableData[0].text()).toBe("01");
    expect(tableData[1].text()).toBe("alert1");
    expect(tableData[2].text()).toContain("logs");
    expect(tableData[3].text()).toBe("default");
    expect(tableData[4].text()).toBe("--");
    expect(tableData[5].text()).toBe("code EqualTo 500");
    expect(tableData[6].text()).toBe("dest1");
  });

  it("Should display add alert button", () => {
    expect(wrapper.find('[data-test="alert-list-add-alert-btn"]').text()).toBe(
      "Add alert"
    );
  });
  it("Should move to add alerts page on clicking on add alert", async () => {
    const routerPush = vi.spyOn(router, "push");
    await wrapper
      .find('[data-test="alert-list-add-alert-btn"]')
      .trigger("click");
    expect(routerPush).toHaveBeenCalledTimes(1);
    expect(routerPush).toHaveBeenCalledWith({
      name: "alertList",
      query: {
        action: "add",
        org_identifier: "default_organization_01",
      },
    });
  });

  describe("When user clicks on edit alert", () => {
    const stream_name = "default";
    const alert_name = "alert1";
    const deleteAlert = vi.spyOn(AlertService, "delete");
    beforeEach(async () => {
      global.server.use(
        rest.delete(
          `${store.state.API_ENDPOINT}/api/${store.state.selectedOrganization.identifier}/${stream_name}/alerts/${alert_name}`,
          (req: any, res: any, ctx: any) => {
            return res(ctx.status(200), ctx.json({ code: 200 }));
          }
        )
      );
      global.server.use(
        rest.get(
          `${store.state.API_ENDPOINT}/api/${store.state.selectedOrganization.identifier}/alerts`,
          (req: any, res: any, ctx: any) => {
            return res(
              ctx.status(200),
              ctx.json({
                list: [
                  {
                    name: "alert2",
                    stream: "default",
                    query: {
                      sql: "select * from  'default'",
                      from: 0,
                      size: 100,
                      start_time: 0,
                      end_time: 0,
                      sql_mode: "full",
                      query_type: "",
                      track_total_hits: false,
                    },
                    condition: {
                      column: "code",
                      operator: "EqualTo",
                      ignore_case: null,
                      value: 500,
                    },
                    duration: 1,
                    frequency: 2,
                    time_between_alerts: 2,
                    destination: "dest1",
                    is_real_time: false,
                  },
                ],
              })
            );
          }
        )
      );
      await wrapper
        .find(`[data-test=alert-list-${alert_name}-delete-alert]`)
        .trigger("click");
      const mainWrapper = new DOMWrapper(document.body);
      await mainWrapper.find('[data-test="confirm-button"]').trigger("click");
      await flushPromises();
    });

    it("Should delete alert from the list", () => {
      expect(deleteAlert).toHaveBeenCalledTimes(1);
    });
    it("Should refetch all alerts", () => {
      const tableRows = wrapper
        .find('[data-test="alert-list-table"]')
        .find("thead")
        .findAll("tr");
      expect(tableRows.length).toBe(1);
      expect(tableRows[0].html()).not.toContain(alert_name);
    });
  });
});
