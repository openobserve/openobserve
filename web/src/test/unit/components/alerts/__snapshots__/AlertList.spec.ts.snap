// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert List > Should match previous table snapshot 1`] = `
"<table class=\\"q-table\\">
  <thead>
    <tr>
      <th class=\\"text-left\\">#</th>
      <th class=\\"text-left sortable\\">Name<i class=\\"q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left\\" aria-hidden=\\"true\\" role=\\"presentation\\">arrow_upward</i></th>
      <th class=\\"text-left sortable\\">Stream Name<i class=\\"q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left\\" aria-hidden=\\"true\\" role=\\"presentation\\">arrow_upward</i></th>
      <th class=\\"text-left sortable\\">Query<i class=\\"q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left\\" aria-hidden=\\"true\\" role=\\"presentation\\">arrow_upward</i></th>
      <th class=\\"text-left sortable\\">Condition<i class=\\"q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left\\" aria-hidden=\\"true\\" role=\\"presentation\\">arrow_upward</i></th>
      <th class=\\"text-left sortable\\">Destination<i class=\\"q-icon notranslate material-icons q-table__sort-icon q-table__sort-icon--left\\" aria-hidden=\\"true\\" role=\\"presentation\\">arrow_upward</i></th>
      <th class=\\"text-center\\">Actions</th>
    </tr>
  </thead>
  <tbody>
    <tr class=\\"\\">
      <td class=\\"text-left\\">01</td>
      <td class=\\"text-left\\">alert1</td>
      <td class=\\"text-left\\">default</td>
      <td class=\\"text-left\\" style=\\"width: 30vw; word-break: break-all;\\">--</td>
      <td class=\\"text-left\\" style=\\"width: 10vw; word-break: break-all;\\">code EqualTo 500</td>
      <td class=\\"text-left\\">dest1</td>
      <td class=\\"q-td text-center\\"><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-ml-xs iconHoverBtn\\" style=\\"font-size: 10px; padding: 8px; min-width: 0; min-height: 0;\\" tabindex=\\"0\\" type=\\"button\\" title=\\"Edit\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">edit</i></span></button><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-ml-xs iconHoverBtn\\" style=\\"font-size: 10px; padding: 8px; min-width: 0; min-height: 0;\\" tabindex=\\"0\\" type=\\"button\\" title=\\"Delete\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><span class=\\"q-icon\\" aria-hidden=\\"true\\" role=\\"img\\"><img src=\\"/src/assets/images/common/delete_icon.svg\\"></span></span></button></td>
    </tr>
    <tr class=\\"\\">
      <td class=\\"text-left\\">02</td>
      <td class=\\"text-left\\">alert2</td>
      <td class=\\"text-left\\">default</td>
      <td class=\\"text-left\\" style=\\"width: 30vw; word-break: break-all;\\">select * from 'default'</td>
      <td class=\\"text-left\\" style=\\"width: 10vw; word-break: break-all;\\">code EqualTo 500</td>
      <td class=\\"text-left\\">dest1</td>
      <td class=\\"q-td text-center\\"><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-ml-xs iconHoverBtn\\" style=\\"font-size: 10px; padding: 8px; min-width: 0; min-height: 0;\\" tabindex=\\"0\\" type=\\"button\\" title=\\"Edit\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">edit</i></span></button><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--round q-btn--actionable q-focusable q-hoverable q-ml-xs iconHoverBtn\\" style=\\"font-size: 10px; padding: 8px; min-width: 0; min-height: 0;\\" tabindex=\\"0\\" type=\\"button\\" title=\\"Delete\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><span class=\\"q-icon\\" aria-hidden=\\"true\\" role=\\"img\\"><img src=\\"/src/assets/images/common/delete_icon.svg\\"></span></span></button></td>
    </tr>
  </tbody>
</table>"
`;
